external-crates/move/crates/move-package/src/resolution/dependency_graph.rs

```diff
@@ -355,6 +359,132 @@ impl<Progress: Write> DependencyGraphBuilder<Progress> {
+                            format!(
+                                "Parsing response from '{resolver}' for dependency group \
+                            '{EXTERNAL_RESOLVER_DEPENDENCY_GROUP}' of package '{parent_pkg_id}'"
+                            )
+                        })?;
+
+                        let pkg_graph = DependencyGraphInfo::new(
+                            sub_graph.clone(),
+                            mode,
+                            /* is_override */ false,
+                            /* is_external */ true,
+                            /* resolved_version */ None,
+                        );
+
+                        // Resolve the root package of this subgraph so that we have a dependency id
+                        // to identify it by. It is expected that if a resolver processes the reserved `_`
+                        // dependency, then that resolver knows to insert a root package ID in the lock
+                        // file contents corresponding to the package itself. Why?
+                        //
+                        // Consider that the external resolver code is normally set up to have knowledge of
+                        // a root package ID whose subgraph it will add for resolution: it is specified in the
+                        // Move.toml as in:
+                        //
+                        // [dependencies]
+                        // Foo = { resolver = "...", ... }
+                        //
+                        // And here we may use the fact that we see ID `Foo` to identify the package's root of the subgraph.
+                        // This is not the case when the resolver processes a dependency group `_` because there is no
+                        // foreknowledge of a legitimate package ID: we have only names or aliases, and a legitimate
+                        // package ID remains to be resolved by the resolver. The mechanism by which the resolver
+                        // communicates the root package ID of a dependency is by returning it in the lock contents
+                        // (subgraph) itself, where it is the root dependency.
+                        let mut root_sub_package = None;
+                        for (_from_package, to_pkg, _) in
+                            sub_graph.package_graph.edges(parent_pkg_id)
+                        {
+                            if root_sub_package.is_some() {
+                                bail!(
+                                    "Multiple root packages in external dependencies found but expected only one",
+                                );
+                            }
+                            root_sub_package = Some(to_pkg);
+                        }
+                        let Some(root_sub_package) = root_sub_package else {
+                            bail!("Expected a root package in external dependencies group but none found");
+                        };
```

Comment:
This was originally a PR comment but I'm erring on the side of verbosity. The comment explains the need for, and tradeoff in the current design where everything is communicated via lock contents. The current schema is designed to expect to know a root package ID of an externally dependency by reading the `Move.toml` of the package. There is no way to specify that when externally resolving a group of dependencies though (and by the nature of the problem, we want the external resolver to determine those package IDs, because we don't have foreknowledge of them).

I've tested and convinced myself that the resolver can set up the Move.lock contents in such a way that we can identify the dependency's root package ID by having it be inserted in the Move.lock contents. Then, this resolver can detect the root package ID given by the graph structure.

However, if there are concerns, we can plausibly change the format that a resolver should output to something more structured (JSON, or whatever) that separates "Root ID package resolved" and "Lock file contents". I'd like to think that we can change that as needed and that it's not blocking, but curious to hear from reviewers.